server/src/main/java/org/elasticsearch/repositories/ShardGenerations.java

```diff
@@ -0,0 +1,168 @@
+     * @param shardId Shard Id
+     * @return generation of the {@link org.elasticsearch.index.snapshots.blobstore.BlobStoreIndexShardSnapshots} blob
+     */
+    public String getShardGen(IndexId indexId, int shardId) {
+        final List<String> generations = shardGenerations.get(indexId);
+        if (generations == null || generations.size() < shardId + 1) {
+            return null;
+        }
+        return generations.get(shardId);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(RepositoryData.SHARDS);
+        for (Map.Entry<IndexId, List<String>> entry : shardGenerations.entrySet()) {
+            builder.array(entry.getKey().getId(), entry.getValue().toArray(Strings.EMPTY_ARRAY));
+        }
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (this == o) {
+            return true;
+        }
+        if (o == null || getClass() != o.getClass()) {
+            return false;
+        }
+        final ShardGenerations that = (ShardGenerations) o;
+        return shardGenerations.equals(that.shardGenerations);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(shardGenerations);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static final class Builder {
+
+        private final Map<IndexId, Map<Integer, String>> raw = new HashMap<>();
+
+        public Builder filterByIndices(Set<IndexId> indices) {
+            raw.keySet().retainAll(indices);
+            return this;
+        }
+
+        public Builder add(ShardGenerations shardGenerations) {
+            shardGenerations.shardGenerations.forEach((indexId, gens) -> {
+                for (int i = 0; i < gens.size(); i++) {
+                    final String gen = gens.get(i);
+                    if (gen != null) {
+                        add(indexId, i, gens.get(i));
+                    }
+                }
+            });
+            return this;
+        }
+
+        public Builder add(IndexId indexId, int shardId, String generation) {
+            raw.computeIfAbsent(indexId, i -> new HashMap<>()).put(shardId, generation);
+            return this;
+        }
+
+        public ShardGenerations build() {
+            return new ShardGenerations(raw.entrySet().stream().collect(Collectors.toMap(
+                Map.Entry::getKey,
+                entry -> {
+                    final Set<Integer> shardIds = entry.getValue().keySet();
+                    assert shardIds.isEmpty() == false;
+                    final int size = shardIds.stream().mapToInt(i -> i).max().getAsInt() + 1;
+                    final String[] gens = new String[size];
+                    entry.getValue().forEach((shardId, generation) -> gens[shardId] = generation);
+                    return Arrays.asList(gens);
```

Comment:
Yes, if a shard failed to update then `null` is used to signifie no change (or none exists yet, try to recover from numeric, if the previous generations had a `null` entry as well either due to being from a previous version or because the repo/index is new).

Reply:
Sure done :)